### PR TITLE
Feature/datadog ecs

### PR DESCRIPTION
**Pull request by DhrubajyotiSadhu**
### :pencil: Description
-  Added datadog to ECS clusters.

### :link: Related Issues
-

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,10 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## [4.1.6] - 2023-01-02
```
Comment (JayGreeeen) [inline CHANGELOG.md:6]: You're a year behind!

Comment (DhrubajyotiSadhu) [reply CHANGELOG.md]: done :D

File: templates/waggledance.json [hunk 2]
```diff
@@ -50,7 +55,11 @@
 	  {
 		"name": "LOG4J_FORMAT_MSG_NO_LOOKUPS",
 		"value": "true"
-	  }
+	  },
+	{
```
Comment (abhimanyugupta07) [inline templates/waggledance.json:59]: formatting needed. Align it like the other blocks above.

File: templates/waggledance.json [hunk 3]
```diff
@@ -84,6 +93,36 @@
                 "namespace": "net.ipv4.tcp_keepalive_probes",
                 "value": "${tcp_keepalive_probes}"
           }
-        ] 
-  }
+        ]
+  },
```
Comment (patduin) [inline templates/waggledance.json:97]: Check the other template files for example but this should be optional.

File: templates/waggledance.json [hunk 4]
```diff
@@ -50,7 +55,11 @@
 	  {
 		"name": "LOG4J_FORMAT_MSG_NO_LOOKUPS",
 		"value": "true"
-	  }
+	  },
+	{
+		"name": "DD_USE_DOGSTATSD",
```
Comment (abhimanyugupta07) [inline templates/waggledance.json:60]: does this need to be parameterised here in the module  ?

File: common.tf [hunk 5]
```diff
@@ -48,3 +48,24 @@ data "aws_iam_policy_document" "waggle_dance_glue_policy" {
...
+data "aws_secretsmanager_secret_version" "datadog_key" {
+  secret_id = data.aws_secretsmanager_secret.datadog_key.id
+}
+
+provider "datadog" {
+  api_key  = jsondecode(data.aws_secretsmanager_secret_version.datadog_key.secret_string).api_key
+  app_key  = jsondecode(data.aws_secretsmanager_secret_version.datadog_key.secret_string).app_key
+}
+
+output "datadog_api_key" {
```
Comment (patduin) [inline common.tf:65]: we don't need these

File: common.tf [hunk 6]
```diff
@@ -48,3 +48,24 @@ data "aws_iam_policy_document" "waggle_dance_glue_policy" {
     ]
   }
 }
+
+data "aws_secretsmanager_secret" "datadog_key" {
+  name  = var.datadog_key_secret_name
+}
+
+data "aws_secretsmanager_secret_version" "datadog_key" {
+  secret_id = data.aws_secretsmanager_secret.datadog_key.id
```
Comment (abhimanyugupta07) [inline common.tf:57]: `datadog_key_secret_name` is set as "" by default. what is the behaviour of this code when a valid secret is not created in the secrets manager ? 

Does it fail the deployment ? I think we want to make this an optional feature.

File: variables.tf [hunk 7]
```diff
@@ -376,4 +376,16 @@ variable "tcp_keepalive_probes" {
   description = "Sets net.ipv4.tcp_keepalive_probes (number), currently only supported in ECS."
   type        = number
   default     = 2
-}
\ No newline at end of file
+}
+
+variable "datadog_key_secret_name" {
```
Comment (abhimanyugupta07) [inline variables.tf:381]: new variables need to be added to README. Please add in the variable's description that this needs to be created in AWS secrets manager.

File: templates/waggledance.json (hunk 4 above)
Comment (DhrubajyotiSadhu) [reply templates/waggledance.json:59]: no this is set to true by default. but will be used only when datadog-agent is running.

File: templates/waggledance.json (hunk 2 above)
Comment (DhrubajyotiSadhu) [reply templates/waggledance.json:58]: done

File: variables.tf (hunk 7 above)
Comment (DhrubajyotiSadhu) [reply variables.tf:381]: Added them in README. Thanks

File: README.md [hunk 8]
```diff
@@ -4,59 +4,62 @@
...
+| datadog_key_secret_name            | SecretsManager secret name containing a DataDog API key. This needs to be created manually in AWS secrets manager.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | number |     `200`      | no |
```
Comment (abhimanyugupta07) [inline README.md:60]:
```suggestion
| datadog_key_secret_name            | Name of the secret containing the DataDog API key. This needs to be created manually in AWS secrets manager.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | number |     `200`      | no |
```

File: templates/waggledance.json (hunk 2 above)
Comment (abhimanyugupta07) [reply templates/waggledance.json:58]: I don't see it fixed - https://github.com/ExpediaGroup/apiary-federation/blob/76f8ab5919b1ce7a4e006bba3c7f7cd005c79546/templates/waggledance.json

File: templates/waggledance.json (hunk 4 above)
Comment (abhimanyugupta07) [reply templates/waggledance.json:59]: what is the use of this variable ? what happens if agent is running and this is set to false?

File: variables.tf [hunk 9]
```diff
@@ -376,4 +376,22 @@ variable "tcp_keepalive_probes" {
...
+}
+
+variable "datadog_agent_version" {
+  description = "Version of the Datadog Agent running in the ECS cluster"
+  type        = string
+  default     = "7.46.0-jmx"
+}
+
+variable "include_datadog_agent" {
+  description = "Whether to include the datadog-agent container alongside waggledance"
```
Comment (abhimanyugupta07) [inline variables.tf:394]:
```suggestion
  description = "Whether to include the datadog-agent container alongside waggledance. This is only applicable to ECS deployments."
```

File: variables.tf [hunk 10]
```diff
@@ -376,4 +376,22 @@ variable "tcp_keepalive_probes" {
...
+}
+
+variable "datadog_key_secret_name" {
+  description = "SecretsManager secret name containing a DataDog API key"
+  type        = string
+  default     = ""
+}
+
+variable "datadog_agent_version" {
+  description = "Version of the Datadog Agent running in the ECS cluster"
```
Comment (abhimanyugupta07) [inline variables.tf:388]:
```suggestion
  description = "Version of the Datadog Agent running in the ECS cluster. This is only applicable to ECS deployments."
```

File: variables.tf [hunk 11]
```diff
@@ -376,4 +376,22 @@ variable "tcp_keepalive_probes" {
   description = "Sets net.ipv4.tcp_keepalive_probes (number), currently only supported in ECS."
   type        = number
   default     = 2
-}
\ No newline at end of file
+}
+
+variable "datadog_key_secret_name" {
+  description = "SecretsManager secret name containing a DataDog API key"
```
Comment (abhimanyugupta07) [inline variables.tf:382]:
```suggestion
  description = "Name of the secret containing the DataDog API key. This needs to be created manually in AWS secrets manager."
```

**Review by abhimanyugupta07 — CHANGES_REQUESTED** [review]
To generate the markdown table for README. Update the variable description and then use 

```
terraform-docs markdown table
```

https://github.com/terraform-docs/terraform-docs

File: variables.tf [hunk 12]
```diff
@@ -348,7 +348,7 @@ variable "datadog_metrics_waggledance" {
   ]
 }
 
-variable "datadog_metrics_port" {
+variable "metrics_port" {
```
Comment (abhimanyugupta07) [inline variables.tf:351]: this variable is not in README. Please regenerate the markdown table using the terraform-docs command.

File: CHANGELOG.md [hunk 13]
```diff
@@ -3,6 +3,10 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## [4.1.6] - 2024-01-03
```
Comment (patduin) [inline CHANGELOG.md:6]: new feature? 4.2.0

**Review by patduin — CHANGES_REQUESTED** [review]
looks ok, @abhimanyugupta07 has some open comments and let's bump the minor version.

**Review by patduin — DISMISSED** [review]

File: templates/waggledance.json (hunk 4 above)
Comment (DhrubajyotiSadhu) [reply templates/waggledance.json]: Removed it as the default value is true.

File: common.tf (hunk 6 above)
Comment (DhrubajyotiSadhu) [reply common.tf]: made it optional.

**Review by patduin — DISMISSED** [review]

**Review by abhimanyugupta07 — APPROVED** [review]

**Review by abhimanyugupta07 — APPROVED** [review]

**Review by patduin — APPROVED** [review]

**Pull request merged**